Add quick blog post about AudioContext outputLatency (#7991)
GoogleChrome/web.dev: src/site/content/en/blog/audio-output-latency/index.md

---
title: Synchronize audio and video playback on the web
subhead: >
  A new Web Audio feature makes it possible to properly achieve AV-sync.
date: 2022-05-25
hero: image/vvhSqZboQoZZN9wBvoXq72wzGAf1/9nYDoT45VHvNAtTfLciY.jpeg

> alt: Flat screen TV turn on inside room photo

Reviewer: This looks broken, not sure what you wanted to say.

authors: 
  - beaufortfrancois
description: >
  A new Web Audio feature makes it possible to properly achieve AV-sync.
tags:
  - blog
  - audio
  - media
---

> The `outputLatency` property of an `AudioContext` instance provides an estimate of audio hardware’s output latency (e.g. Bluetooth earbuds or USB external audio interface). This property is necessary to coordinate/compensate the latency from the input to the output. And it is also very useful when you need to synchronize a video stream and an audio stream produced by the [Web Audio API](https://developer.mozilla.org/docs/Web/API/Web_Audio_API). The process of synchronizing video and audio streams together is called [AV-sync](https://en.wikipedia.org/wiki/Audio-to-video_synchronization).

Reviewer: ```suggestion
The `outputLatency` property of an `AudioContext` instance provides an estimate of audio hardware's output latency (e.g. that of Bluetooth earbuds or of an external USB audio interface). This property is necessary to coordinate/compensate the latency from the input to the output. And it is also very useful when you need to synchronize a video stream and an audio stream produced by the [Web Audio API](https://developer.mozilla.org/docs/Web/API/Web_Audio_API). The process of synchronizing video and audio streams together is called [AV-sync](https://en.wikipedia.org/wiki/Audio-to-video_synchronization).
```

> The `outputLatency` property of an `AudioContext` instance provides an estimate of audio hardware's output latency (e.g. that of Bluetooth earbuds or of an external USB audio interface). This property is necessary to coordinate/compensate the latency from the input to the output. And it is also very useful when you need to synchronize a video stream and an audio stream produced by the [Web Audio API](https://developer.mozilla.org/docs/Web/API/Web_Audio_API). The process of synchronizing video and audio streams together is called [AV-sync](https://en.wikipedia.org/wiki/Audio-to-video_synchronization).

Reviewer 2: ```suggestion
The `outputLatency` property of an `AudioContext` instance provides an estimate of audio hardware's output latency (for example, that of Bluetooth earbuds or of an external USB audio interface). This property is necessary to coordinate/compensate the latency from the input to the output. And it is also very useful when you need to synchronize a video stream and an audio stream produced by the [Web Audio API](https://developer.mozilla.org/docs/Web/API/Web_Audio_API). The process of synchronizing video and audio streams together is called [AV-sync](https://en.wikipedia.org/wiki/Audio-to-video_synchronization).
```

> In [Chris Cunningham's demo](https://github.com/chcunningham/wc-talk#simple_video_playerhtml-live-demo), [WebCodecs](https://developer.mozilla.org/docs/Web/API/WebCodecs_API) is used to decode a `MediaStream` into raw video and audio data, and then played back into a HTML canvas element with audio data coming from the Web Audio API. The `outputLatency` property allows the demo to determine when a given audio timestamp is reaching the user's ears and then properly paint video frames to match that.

Reviewer 2: ```suggestion
In [Chris Cunningham's demo](https://github.com/chcunningham/wc-talk#simple_video_playerhtml-live-demo), the [WebCodecs API](https://developer.mozilla.org/docs/Web/API/WebCodecs_API) is used to decode a `MediaStream` into raw video and audio data, and then played back into a HTML canvas element with audio data coming from the Web Audio API. The `outputLatency` property allows the demo to determine when a given audio timestamp is reaching the user's ears and then properly paint video frames to match that.
```

> {% Img src="image/vvhSqZboQoZZN9wBvoXq72wzGAf1/p61K2S4Wye34OJn4XZOj.png", alt="A screenshot of the demo", width="800", height="533" %}

Reviewer 2: ```suggestion
<figure>
  {% Img src="image/vvhSqZboQoZZN9wBvoXq72wzGAf1/p61K2S4Wye34OJn4XZOj.png", alt="A screenshot of the demo", width="800", height="533" %}
</figure>
```

> Try it out for yourself, play the video with your favorite Bluetooth headset 🎧, wait for the bird 🐦 (see above), and toggle the checkbox ☑️ to observe audio playback changes. The total output latency value is updated in real-time.

Reviewer 2: ```suggestion
Try it out for yourself, play the video with your favorite Bluetooth headset (🎧), wait for the bird (🐦) (see above), and toggle the checkbox (☑️) to observe audio playback changes. The total output latency value is updated in real-time.
```

### AudioContext outputLatency

{% BrowserCompat 'api.AudioContext.outputLatency' %}

> Hero image by [Wahid Khene](https://unsplash.com/@wahidkhene) on [Unsplash](https://unsplash.com/photos/iKdQCIiSMlQ).

Reviewer 2: ```suggestion
_Hero image by [Wahid Khene](https://unsplash.com/@wahidkhene) on [Unsplash](https://unsplash.com/photos/iKdQCIiSMlQ)._
```

---
title: Synchronize audio and video playback on the web
subhead: >

>   A new Web Audio feature makes it possible to properly achieve AV-sync.

Author: I've addressed your feedback. Good catch!

>   A new Web Audio feature makes it possible to properly achieve AV-sync.

Reviewer 3: Correct. This is not new. Chrome is catching up the spec.

date: 2022-05-25
hero: image/vvhSqZboQoZZN9wBvoXq72wzGAf1/9nYDoT45VHvNAtTfLciY.jpeg
alt: A photo of a video editing timeline in Adobe Premiere Pro
authors: 
  - beaufortfrancois
description: >
  A new Web Audio feature makes it possible to properly achieve AV-sync.
tags:
  - blog
  - audio
  - media
---

> The `outputLatency` property of an `AudioContext` instance provides an estimate of audio hardware's output latency (for example, that of Bluetooth earbuds or of an external USB audio interface). This property is necessary to coordinate/compensate the latency from the input to the output. And it is also very useful when you need to synchronize a video stream and an audio stream produced by the [Web Audio API](https://developer.mozilla.org/docs/Web/API/Web_Audio_API). The process of synchronizing video and audio streams together is called [AV synchronization](https://en.wikipedia.org/wiki/Audio-to-video_synchronization).

Reviewer 3: > This property is necessary to coordinate/compensate the latency from the input to the output.

Can we clarify this? The output latency can be useful when you want to:
- Synchronize the existing audio material and the newly recorded material. (in a music production scenario)
- Synchronize the Web Audio output and other media (e.g. video or MIDI playback).

So "coordinating from the input to the output" doesn't seem clear enough on what it does. Perhaps I am missing something.

> In [Chris Cunningham's demo](https://github.com/chcunningham/wc-talk#simple_video_playerhtml-live-demo), the [WebCodecs API](https://developer.mozilla.org/docs/Web/API/WebCodecs_API) is used to decode a `MediaStream` into raw video and audio data, and then played back into a HTML `<canvas>` element with audio data coming from the Web Audio API. The `outputLatency` property allows the demo to determine when a given audio timestamp is reaching the user's ears and then properly paint video frames to match that.

Reviewer 3: I was hoping to build a similar example under GoogleChromeLabs so we can have it as a reference implementation. But we can update the article when it is ready.

---
title: Synchronize audio and video playback on the web
subhead: >
  The Web Audio API makes it possible to properly achieve AV synchronization.

> date: 2022-05-25

Reviewer 2: ```suggestion
date: 2022-05-27
```

tags:
  - blog
  - audio
  - media
---

The `outputLatency` property of an `AudioContext` instance provides an estimate of audio hardware's output latency (for example, that of Bluetooth earbuds or of an external USB audio interface). This property is useful when you want to:

- Synchronize the existing audio material and the newly recorded material. (in a music production scenario)
- Synchronize the [Web Audio](https://developer.mozilla.org/docs/Web/API/Web_Audio_API) output and other media (e.g. video or MIDI playback).

In this [WebCodecs demo](https://wc-talk.netlify.app/simple_video_player.html) ([source](https://github.com/chcunningham/wc-talk)), the [WebCodecs API](https://developer.mozilla.org/docs/Web/API/WebCodecs_API) is used to decode a `MediaStream` into raw video and audio data, and then played back into a HTML `<canvas>` element with audio data coming from an [Audio Worklet](https://developer.mozilla.org/docs/Web/API/AudioWorklet). The `outputLatency` property allows the demo to determine when a given audio timestamp is reaching the user's ears and then properly paint video frames to match that.

<figure>

>   {% Img src="image/vvhSqZboQoZZN9wBvoXq72wzGAf1/p61K2S4Wye34OJn4XZOj.png", alt="A screenshot of the demo", width="800", height="533" %}

Reviewer 2: ```suggestion
  {% Img src="image/vvhSqZboQoZZN9wBvoXq72wzGAf1/p61K2S4Wye34OJn4XZOj.png", alt="A screenshot of the demo, which is of an embedded video with audio controls for volume, audio buffer health, total output latency, and a checkbox to use AudioContext.outputLatency.", width="800", height="533" %}
```

---
title: Synchronize audio and video playback on the web
subhead: >
  The Web Audio API makes it possible to properly achieve AV synchronization.
date: 2022-05-25
hero: image/vvhSqZboQoZZN9wBvoXq72wzGAf1/9nYDoT45VHvNAtTfLciY.jpeg
alt: A photo of a video editing timeline in Adobe Premiere Pro
authors: 
  - beaufortfrancois
description: >
  The Web Audio API makes it possible to properly achieve AV synchronization.
tags:
  - blog
  - audio
  - media
---

The `outputLatency` property of an `AudioContext` instance provides an estimate of audio hardware's output latency (for example, that of Bluetooth earbuds or of an external USB audio interface). This property is useful when you want to:

- Synchronize the existing audio material and the newly recorded material. (in a music production scenario)
- Synchronize the [Web Audio](https://developer.mozilla.org/docs/Web/API/Web_Audio_API) output and other media (e.g. video or MIDI playback).

> The process of synchronizing video and audio streams together is called [AV synchronization](https://en.wikipedia.org/wiki/Audio-to-video_synchronization).

Author: I've removed it.
